### PR TITLE
Predictive GFM 2025

### PR DESCRIPTION
**Pull request by allaffa**
Main changes: 

1. Consistency check across all large scale datasets for the format of the torch.geometric.data objects

 ```
               data_object = Data(
                    dataset_name="dataset_name",
                    natoms=natoms,
                    pos=pos,
                    cell=None,  # even if not needed, cell needs to be defined because ADIOS requires consistency across datasets
                    pbc=None,  # even if not needed, pbc needs to be defined because ADIOS requires consistency across datasets
                    edge_shifts=None,  # even if not needed, edge_shift needs to be defined because ADIOS requires consistency across datasets
                    atomic_numbers=atomic_numbers,  # Reshaping atomic_numbers to Nx1 tensor
                    chemical_composition=chemical_composition,
                    smiles_string=smiles_string,
                    x=x,
                    energy=energy,
                    energy_per_atom=energy_per_atom,
                    force=forces,
                )
```

2. apply `graphgps_transform` to compute structural and positional Laplacian encodings
3. Allow for parsed input argument choice to set `compute_grad_energy`. Default values is `False`
4. default value of `energy_per_atom` is set to `False`, because we do not need to normalize for machine learning force fields
5. Added chemical composition as one-dimensional vector with 118 entries. each entry counts the number of atoms for that chemical species in the atomistic structures

### CODE REVIEW

File: examples/ani1_x/train.py [hunk 1]
```diff
@@ -103,35 +113,56 @@ def convert_trajectories_to_graphs(self):
...
+                data_object = Data(
+                    dataset_name="ani1x",
                     natoms=natoms,
-                    # stress=torch.tensor(stresses, dtype=torch.float32),
-                    # magmom=torch.tensor(magmom, dtype=torch.float32),
-                    pos=positions,
+                    pos=pos,
+                    cell=None,  # even if not needed, cell needs to be defined because ADIOS requires consistency across datasets
+                    pbc=None,  # even if not needed, pbc needs to be defined because ADIOS requires consistency across datasets
+                    edge_shifts=None,  # even if not needed, edge_shift needs to be defined because ADIOS requires consistency across datasets
```
Comment (RylieWeaver) [inline examples/ani1_x/train.py:141]: @allaffa 

I think you will want to make a dummy edge_attr here for consistency for consistency for ADIOS. Either that, or comment the three lines in RadiusGraphPBC that create a data.edge_attr from distance.

File: examples/qm7x/train.py [hunk 2]
```diff
@@ -202,23 +214,65 @@ def hdf5_to_graph(self, fMOL, molid):
...
+                hist, _ = np.histogram(atomic_number_list, bins=range(1, 118 + 2))
+                chemical_composition = torch.tensor(hist).unsqueeze(1).to(torch.float32)
+
+                data_object = Data(
+                    dataset_name="qm7x",
+                    natoms=natoms,
+                    pos=pos,
+                    cell=None,  # even if not needed, cell needs to be defined because ADIOS requires consistency across datasets
+                    pbc=None,  # even if not needed, pbc needs to be defined because ADIOS requires consistency across datasets
+                    edge_shifts=None,  # even if not needed, edge_shift needs to be defined because ADIOS requires consistency across datasets
```
Comment (RylieWeaver) [inline examples/qm7x/train.py:235]: @allaffa 

I recommend a dummy edge_attr here as well.

File: examples/ani1_x/train.py (hunk 1 above)
Comment (allaffa) [reply examples/ani1_x/train.py:141]: @RylieWeaver 

The function RadiysGraph or the derivative RadiusGraphPBC (which are used in every example) create edge-index and set edge_attr=None already 

https://pytorch-geometric.readthedocs.io/en/2.6.1/_modules/torch_geometric/transforms/radius_graph.html#RadiusGraph

File: examples/qm7x/train.py (hunk 2 above)
Comment (allaffa) [reply examples/qm7x/train.py:235]: @RylieWeaver 
Please see comment above for `ani1-x`

File: examples/ani1_x/train.py (hunk 1 above)
Comment (RylieWeaver) [reply examples/ani1_x/train.py:143]: Ah, you are right, Radius Graph creates an edge_attr by default as well. 

This should not be a problem.

File: examples/ani1_x/train.py [hunk 3]
```diff
@@ -103,35 +113,58 @@ def convert_trajectories_to_graphs(self):
...
+                    edge_index = None,
+                    edge_attr=None,
+                    edge_shifts=None,  # even if not needed, edge_shift needs to be defined because ADIOS requires consistency across datasets
                     atomic_numbers=atomic_numbers,  # Reshaping atomic_numbers to Nx1 tensor
+                    chemical_composition=chemical_composition,
                     x=x,
-                    y=energy,
+                    energy=energy,
+                    energy_per_atom=energy_per_atom,
+                    force=forces,
```
Comment (RylieWeaver) [inline examples/ani1_x/train.py:149]: @allaffa

This may need to be plural, since `compute_grad_energy` expects the convention `data.forces`

File: examples/qm7x/train.py [hunk 4]
```diff
@@ -202,23 +214,67 @@ def hdf5_to_graph(self, fMOL, molid):
...
+                    pbc=None,  # even if not needed, pbc needs to be defined because ADIOS requires consistency across datasets
+                    edge_index=None,
+                    edge_attr=None,
+                    edge_shifts=None,  # even if not needed, edge_shift needs to be defined because ADIOS requires consistency across datasets
+                    atomic_numbers=atomic_numbers,  # Reshaping atomic_numbers to Nx1 tensor
+                    chemical_composition=chemical_composition,
+                    x=x,
+                    energy=energy,
+                    energy_per_atom=energy_per_atom,
+                    force=forces,
```
Comment (RylieWeaver) [inline examples/qm7x/train.py:243]: @allaffa 

Same here, I think this may need to be plural.

File: examples/ani1_x/train.py (hunk 3 above)
Comment (allaffa) [reply examples/ani1_x/train.py]: @RylieWeaver 

Thanks. Done.

File: examples/qm7x/train.py (hunk 4 above)
Comment (allaffa) [reply examples/qm7x/train.py]: @RylieWeaver 

Thanks. Done.

**Review by RylieWeaver — APPROVED** [review]

**Comment by allaffa** [comment]
@pzhanggit 
This PR corresponds to the branch where we will try to perform our imbalanced, multi-source work. 
Please take a look at the structure of the `Data` objects.

**Comment by allaffa** [comment]
@zachfox 
Whenever we will move ahead with the conditional DM, please take a look at the `Data` structures of this PR

File: examples/omat24/train.py [hunk 5]
```diff
@@ -159,37 +169,63 @@ def __init__(
                     if self.energy_per_atom:
```
Comment (RylieWeaver) [inline examples/omat24/train.py:169]: @allaffa 

I think this could be taken out because of redundant logic. Below it already writes both `energy` and `energy_per_atom`, and assigns `data.y` accordingly.

Comment (RylieWeaver) [reply examples/omat24/train.py:169]: @allaffa 

Other than this small snippet it looks good to me!

**Comment by pzhanggit** [comment]
> @pzhanggit This PR corresponds to the branch where we will try to perform our imbalanced, multi-source work. Please take a look at the structure of the `Data` objects.

Thank you, Max. Introducing a `dataset_name` ID looks good for our multi-source work. Let me know when you complete the datasets generation, and I'll start the multi-source model training.

About the changes in `hydragnn/utils/descriptors_and_embeddings/smiles_utils.py`, I suggest we move them to a standalone file.

**Review by pzhanggit — COMMENTED** [review]
See my replies to Max's comment

**Comment by allaffa** [comment]
> > @pzhanggit This PR corresponds to the branch where we will try to perform our imbalanced, multi-source work. Please take a look at the structure of the `Data` objects.
> 
> Thank you, Max @allaffa . Introducing a `dataset_name` ID looks good for our multi-source work. Let me know when you complete the datasets generation, and I'll start the multi-source model training.
> 
> About the changes in `hydragnn/utils/descriptors_and_embeddings/smiles_utils.py`, I suggest we move them to a standalone file.

@pzhanggit 
Thanks, I moved the functionalities in a separate `xyz2mol.py` script

**Review by pzhanggit — APPROVED** [review]
